### ISSUE TITLE
Content-Disposition of multipart/form-data too strict for some implementations

### ISSUE THREAD

**Issue by arneschuldt** [playframework/playframework#8527]
### Play Version (2.5.x / etc)

2.6.x but also affects master as of today

### API (Scala / Java / Neither / Both)

Both

### Expected Behavior

Play parses the body of POST requests with `Content-Type` `multipart/form-data` in accordance with [RFC 7578](https://tools.ietf.org/html/rfc7578). In particular, the [`Content-Disposition` field for the parts is expected to have the value `form-data`](https://tools.ietf.org/html/rfc7578#section-4.2). The standard mentions no other values for `Content-Disposition` apart from `form-data`.

Preceding (and now outdated) standards [RFC 2388](https://tools.ietf.org/html/rfc2388) (Section 4.4) and [RFC 1867](https://tools.ietf.org/html/rfc1867) (Section 3.3), however, also mention the value `file` for `Content-Disposition`. Note that it is not quite clear if `file` is meant only for nested multipart contents.

Unfortunately, there are countless IoT devices in use which also use `Content-Disposition: file` in addition to `Content-Disposition: form-data`.

The devices of a leading manufacturer of printer/scanner multifunction devices offer a webservice interface that allows specifying a webhook URL to which scanned images should be sent back. In that case, Play acts as the receiver of the POST request.

Expected behaviour is that Play also recognises these files with `Content-Disposition: file` (either by default or by configuration to handle such requests more generously).

(Experiments reveal that, for instance, PHP has this less restrictive behaviour in processing such requests.)

### Actual Behavior

Actual behaviour of Play is that file parts with `Content-Disposition: file` header are not recognised by Play (see test data below).

This is caused by lines 194 and 213 of [`Multipart.scala`](https://github.com/playframework/playframework/blob/c038bc147b03285a1de7acc50037e1e7f438bedf/framework/src/play/src/main/scala/play/core/parsers/Multipart.scala) which require the value `form-data`.

Unfortunately, this prevents Play from interacting with aforementioned scanner devices out of the box.

### Possible Solutions

Updating the scanner firmware is not an option because these devices have been sold for many years worldwide.

However, I suggest the following possible solutions with hopefully minimal changes in Play (I do not provide a PR because I have only very limited knowledge of Scala):

* Change lines 194 and 213 of [`Multipart.scala`](https://github.com/playframework/playframework/blob/c038bc147b03285a1de7acc50037e1e7f438bedf/framework/src/play/src/main/scala/play/core/parsers/Multipart.scala) to `_ <- values.get("name")`. This would completely ignore the `Content-Disposition` type and only check whether there is name (required). This solution would be general in the way that maybe other devices might use [even other types known from other contexts](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition).
* Change lines 194 and 213 to accept both `form-data` and `file`. More restrictive solution than the first one but still solves the specific identified problem.
* Make one of the first two solutions optional with a configuration parameter. Preserves previous behaviour of Play by default (if this is considered necessary for this case).
* Provide a possibilty to inject an own implementation for the type check. Appears to be rather over-engineered.

Unfortunately, I do not see an obvious solution how developers can solve the problem with a minimally invasive modification from Play (Java) projects. I am happy about useful pointers for such an ad-hoc solution we can use while waiting for this issue to be resolved. :)

### Reproducible Test Case

An example POST request by a scanner looks as follows:

```
POST /api/xrx/files HTTP/1.1
Content-Type: multipart/form-data;boundary=--MIMEBoundary633737556018130000
[...]

----MIMEBoundary633737556018130000
Content-Disposition: form-data; name="theOperation"

PutFile
----MIMEBoundary633737556018130000
Content-Disposition: form-data; name="destDir"


----MIMEBoundary633737556018130000
Content-Disposition: form-data; name="destName"

DOC.XSM/NEXTNAME.DAT
----MIMEBoundary633737556018130000
Content-Disposition: file; name="data"; filename="tmp59E3.tmp"
Content-Type: image/tiff
Content-Length: 12
Content-Transfer-Encoding: binary

00000002.tif
----MIMEBoundary633737556018130000--
```

In Play Java, a call to `request().body().asMultipartFormData().asFormUrlEncoded()` lists the first three parts as expected. The last part, however, is not contained in `request().body().asMultipartFormData().getFiles()`.

We have recorded the request, edited it manually (only the `Content-Disposition` of the last part changed from `file` to `form-data`), for a modified replay:

```
POST /api/xrx/files HTTP/1.1
Content-Type: multipart/form-data;boundary=--MIMEBoundary633737556018130000
[...]

----MIMEBoundary633737556018130000
Content-Disposition: form-data; name="theOperation"

PutFile
----MIMEBoundary633737556018130000
Content-Disposition: form-data; name="destDir"


----MIMEBoundary633737556018130000
Content-Disposition: form-data; name="destName"

DOC.XSM/NEXTNAME.DAT
----MIMEBoundary633737556018130000
Content-Disposition: form-data; name="data"; filename="tmp59E3.tmp"
Content-Type: image/tiff
Content-Length: 12
Content-Transfer-Encoding: binary

00000002.tif
----MIMEBoundary633737556018130000--
```

Now, the file is recognised by Play. (In real life this change is not an option because we cannot influence the firmware of the scanner).

**Comment by arneschuldt** [comment]
We have created pull request #8530 implementing solution 1.

**Comment by marcospereira** [comment]
Hi @arneschuldt,

Thanks for the very detailed and informative report.

I think the solution to support `form-data` and `file` is the one we should follow. Although I don't have a strong opinion on this, I prefer to keep the framework as close the RFC as possible (removing such "incompatibilities" in the future is complicated if people depend on them).

So I would go with the second solution proposed.

> In Play Java, a call to request().body().asMultipartFormData().asFormUrlEncoded() lists the first three parts as expected. The last part, however, is not contained in request().body().asMultipartFormData().getFiles().

You can create a body parser to handle this. It may involve to copy and change `Multipart.scala`, but it is a solution while we don't have a release supporting what you described.

**Comment by arneschuldt** [comment]
Hi @marcospereira, thank you for your feedback (and the proposed workaround). I agree with your position and we have updated pull request #8530 accordingly to solution 2.

[MDN mentions two other `Content-Disposition` types in other contexts of HTTP.](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) I would like to mention them here in the issue because they are possible candidates for other non-standard implementations:

* `Content-Disposition: inline`
* `Content-Disposition: attachment`

So, if anyone runs into similar problems in the future, they can easily find this issue and a blueprint for solving it.

**Comment by arneschuldt** [comment]
Original issue solved in pull request #8530, now merged by @marcospereira. Thank you!
